libpod/runtime_migrate.go

```diff
@@ -84,5 +86,112 @@ func (r *Runtime) Migrate(newRuntime string) error {
+		}
+
+		// Recreate the cached paths in the new database
+		if err := sqlState.ValidateDBConfig(r); err != nil {
+			return fmt.Errorf("recreating config table in SQLite: %w", err)
+		}
+
+		// Migrate volumes, then pods, then containers.
+		// Containers must be last as the pods they are part of and volumes they use must already exist.
+		allVolumes, err := r.state.AllVolumes()
+		if err != nil {
+			return fmt.Errorf("retrieving all volumes to migrate to new database: %w", err)
+		}
+		for _, vol := range allVolumes {
+			if err := sqlState.AddVolume(vol); err != nil {
+				return err
+			}
+			if err := vol.update(); err != nil {
+				return err
+			}
+			if err := sqlState.SaveVolume(vol); err != nil {
+				return err
+			}
+		}
+
+		allPods, err := r.state.AllPods()
+		if err != nil {
+			return fmt.Errorf("retrieving all pods to migrate to new database: %w", err)
+		}
+		for _, pod := range allPods {
+			if err := sqlState.AddPod(pod); err != nil {
+				return err
+			}
+			if err := pod.updatePod(); err != nil {
+				return err
+			}
+			if err := sqlState.SavePod(pod); err != nil {
+				return err
+			}
+		}
+
+		// Containers must be done as a graph due to dependencies.
+		// The state will error if we add a container before its dependencies.
+		allCtrs, err := r.state.AllContainers(true)
+		if err != nil {
+			return fmt.Errorf("retrieving all containers to migrate to new database: %w", err)
+		}
+		graph, err := BuildContainerGraph(allCtrs)
+		if err != nil {
+			return err
+		}
+
+		ctrErrors := make(map[string]error)
+		ctrsVisited := make(map[string]bool)
+
+		for _, node := range graph.noDepNodes {
+			migrateNodeDatabase(node, false, ctrErrors, ctrsVisited, sqlState)
+		}
+		var ctrError error
+		for id, err := range ctrErrors {
+			if ctrError != nil {
+				logrus.Errorf("Migrating containers to SQLite: %v", ctrError)
+			}
+			ctrError = fmt.Errorf("migrating container %s: %w", id, err)
+		}
+		if ctrError != nil {
+			return ctrError
+		}
+
+		boltState := r.state
+		r.state = sqlState
+		boltState.Close()
+
+		// Move the Bolt database so it is not reused, but preserve so data is not lost.
+		boltDBPath := getBoltDBPath(r)
+		newBoltDBPath := fmt.Sprintf("%s-old", boltDBPath)
+		if err := os.Rename(boltDBPath, newBoltDBPath); err != nil {
+			return fmt.Errorf("renaming old database %s to %s: %w", boltDBPath, newBoltDBPath, err)
+		}
+		logrus.Errorf("Old database has been renamed to %s and will no longer be used", newBoltDBPath)
+
+		// Warn the user if they explicitly configured BoltDB
+		backend, err := config.ParseDBBackend(r.config.Engine.DBBackend)
+		if err != nil {
+			return fmt.Errorf("invalid DB backend configured - please change containers.conf database_backend to \"sqlite\"")
+		}
+		if backend == config.DBBackendBoltDB {
+			logrus.Errorf("In containers.conf, database_backend is manually set to \"boltdb\" - please change to \"sqlite\" to complete migration")
+		}
```

Comment:
This is wrong as it will always trigger

`runtime.config.Engine.DBBackend = config.DBBackendBoltDB.String()` is set explicitly  in getDBState() when the config file has the default.

I believe you would need to use config.Default(), though even then I am not sure since it might have overwritten the cached default. Maybe a config.New() to actually read the config file values from disk again.